### ISSUE TITLE
aws_s3_object does not detect that original file has changed and re-upload

### ISSUE THREAD

**Issue by timothykimball** [hashicorp/terraform#3068]
Currently you have to manually mark the object tainted and then re-upload.

Using the etag, it should be straightforward for terraform to detect that the object is now different and force a replacement:
```
resource "aws_s3_bucket" "config" {
    bucket = "aire-${var.vpc_name}-config"
    acl = "private"
    force_destroy = true
    tags {
        Name = "A bucket for all your configuration needs"
        Environment = "Dev"
    }
}

resource "aws_s3_bucket_object" "object" {
    bucket = "${aws_s3_bucket.config.id}"
    depends_on = ["aws_s3_bucket.config"]
    key = "cloudwatch/${var.vpc_name}.cfg"
    source = "${path.module}/files/cloudwatch/${var.vpc_name}.cfg"
}
```

In the above example, if the source file changes after uploading the first time, terraform never detects the file has changed subsequent.

**Comment by apparentlymart** [comment]
I think this is more tricky than it first appears, since the ETag is not dependably an MD5 hash of the content when encryption is in use.

However I certainly agree that it's annoying. I wonder if there is a way to set things up so that we only try to do the ETag matching when encryption is not in use, and manual tainting is required only if encryption is enabled.

**Comment by simpsora** [comment]
I've just run into this problem as well.  It would be really useful for Terraform to recognize a changed object and cause it to be re-uploaded.

If you can't reliably use `ETag`, could you keep a local hash of the source file's contents and check that?  I don't know if there's any provision for local metadata in TF or not.

**Comment by simpsora** [comment]
Another option could be the ability to add a lifecycle rule that permanently marks an object as tainted, forcing it to be re-uploaded during every run?  That's not ideal, but IMHO it's better to update files when not necessary rather than not update them when it _is_ necessary.

**Comment by muikrad** [comment]
I have a similar problem but a totally different use case. I am using Terraform to prepare and upload a code package as an AWS Lambda function.

I have a null_resource that launches a script to prepares a zip package, then the lambda resource uploads that file. How can I either:

1. Set the null_resource to always execute.
2. Detect changes in a folder (there are several files in several folders) and repack the zip file as necessary?

It seems the current design expects me to pack the zip file before calling terraform, so I could use the sha256 interpolation on the zip file to detect changes. Manually tainting that resource seems to be the only way around it at the moment.

**Comment by muikrad** [comment]
Spending more time with the use-case, i'm noticing that the sha/file trigger will not work because when the file doesn't exist, the interpolation fails, meaning you can't plan/apply unless the file is there to begin with... :(

**Comment by gregeinfrank** [comment]
+1 any update on this?

**Comment by radeksimko** [comment]
> However I certainly agree that it's annoying. I wonder if there is a way to set things up so that we only try to do the ETag matching when encryption is not in use, and manual tainting is required only if encryption is enabled.

I will confirm this soon, but I think that we're pretty much in control of the encryption since it's happening on the object level - i.e. the resource itself (`aws_s3_bucket_object`) knows whether encryption is or is not used and may even be able to reject invalid configurations.

That said, we don't actually support encryption in `aws_s3_bucket_object` (yet), so it may not be such a problem - we may just postpone that problem to the future.

**Comment by radeksimko** [comment]
Fixed via #5305
